tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -1572,19 +1572,18 @@ Status StorageManager::get_array_schema_uris(
     const URI& array_uri, std::vector<URI>* uris) const {
   STATS_START_TIMER(stats::GlobalStats::TimerType::READ_GET_ARRAY_SCHEMAS)
 
+  // Always add array_name/__array_schema.tdb as the first schema uri
+  uris->clear();
+  auto schema_uri_old = array_uri.join_path(constants::array_schema_filename);
+  uris->emplace_back(schema_uri_old);
+
+  // Add schema uris under directory array_name/__schema
   const URI schema_uri =
       array_uri.join_path(constants::array_schema_folder_name);
-  bool is_dir = false;
-  RETURN_NOT_OK(vfs_->is_dir(schema_uri, &is_dir));
-  if (!is_dir) {
-    // If it is not a directory, the array schema could be stored in the
-    // previous format
-    uris->clear();
-    return Status::Ok();
-  }
-  RETURN_NOT_OK(vfs_->ls(schema_uri, uris));
-  // Sort schema URIs
-  std::sort(uris->begin(), uris->end());
+
+  // Since ls could return NOT Ok status, we will not use RETURN_NOT_OK here
+  vfs_->ls(schema_uri, uris);
```

Comment:
I think I do understand what you are trying to do here, but in verbose mode the error will be printed and will confuse the heck out of the users. We should never let any error be uncaught. We need a different and safer approach here. 

Specifically, I suggest we do the following: 

1. We need to perform a full `ls` in the array directory anyway to get the fragment URIs. We can perform this outside this function, and pass the list to this function as well as of course the function that loads the fragment URIs (we are already doing that). However, it is important to run the `ls` function in the array directory only once when opening the array.
2. After receiving the list of array contents, we need to check for the existence of `<array_name>/__array_schema.tdb` (older version) and `<array_name>/__schema`. This will allow us to know whether we need to perform a second `ls` inside `<array_name>/__schema/` to get the new array schemas. 

Observe that what I am suggesting is optimal in the sense that we perform at most 2 `ls` operations (one in the array folder and one in the `__schema` folder).

Reply:
I will change them in that way.

Reply:
A vector of URI is added as parameter for some related function. Now when we call array_open_for_reads function, we only call ls once to get Uris under <array_name> directory in get_fragment_uris, then pass that vector to get_array_schema_uris.  At the mean time, I only found one function(array_open_for_reads) that can reuse the ls result. If there are more, please let me know.